### ISSUE TITLE
catch unavailable backend

### ISSUE THREAD

**Issue by rsoeldner** [imalsogreg/servant-reflex#98]
Hi,

If the resource is not reachable (ie. server offline) I would have expected a `RequestFailure :: ReqResult` or something. What is the way to handle such cases ?

**Comment by imalsogreg** [comment]
o/

`ReqResult` can be either `ResponseSuccess`, `ResponseFailure`, or `RequestFailure`. The second is returned on cases like service unreachable, the third is returned for malformed requests.

I think the easiest way to handle them is to call `fmapMaybe`[`reqFailure`](https://github.com/imalsogreg/servant-reflex/blob/master/src/Servant/Common/Req.hs#L98-L101) on the result of the client call:

```
  -- ... continued ...
  res :: Event t (ReqResult Double) <- doubleIt xs triggers
  let 
    errs = fmapMaybe reqFailure res
    
  -- Red <p> tag showing the last error
  elAttr "p" ("style" =: "color:red") $
    dynText =<< holdDyn "" errs
```

`reqFailure` doesn't differentiate between request errors and response errors though. If you want your error-handling logic to apply specifically to response errors, you could manually pattern match on `ReqResult`:

```
  -- ... continued ...
  res :: Event t (ReqResult Double) <- doubleIt xs triggers
  let 
    errs503 :: Event t ()
    errs503 = fforMaybe res $ \case
        ResponseError reason xhrResponse
           | _xhrResponse_status == 503 -> Just ()
        _ -> Nothing
    
  -- Red <p> tag shows up when we encounter a 503 error
  elAttr "p" ("style" =: "color:red") $
    dynText =<< holdDyn "" ("Service Unavailable" <$ errs503)
```

Does that answer your question? Or, are you finding that servant-reflex is failing to return an error when the server's offline in your case?

**Comment by rsoeldner** [comment]
yes, this is what I have expected, but unfortunately I see this (WebkitGtk):

I'm using 
```
file:///home/rsoeldner/abc/frontend/index.html:247:76: CONSOLE WARN ArrayBuffer is deprecated in XMLHttpRequest.send(). Use ArrayBufferView instead.
file:///home/rsoeldner/abc/frontend/index.html:386:20: CONSOLE LOG h$wrapBuffer: not an ArrayBuffer
A JavaScript exception was thrown! (may not reach Haskell code)
h$wrapBuffer: not an ArrayBuffer
<interactive>: JSException
```
I'm currently experimenting with servant==0.17 ... I will further check - thank you

**Comment by imalsogreg** [comment]
Ahh, thanks! If you are set up for it, could you check to see whether this happens with a ghcjs build of the frontend too?

**Comment by rsoeldner** [comment]
ghcjs build works ... do you have any idea why the webkit version fails ?

**Comment by imalsogreg** [comment]
I don't know, but it sounds like the underlying XHR handling may not work under webkitgtk. Your error message mentions ArrayBuffer.. I haven't looked into the proper way of receiving response data, but I have a vague memory that there are options at runtime for which type of container to use...

Let's use this issue to track the fix. Could you update the title to indicate that XHR is broken in webkit builds? Is this something you need fixed soon?

**Comment by santiweight** [comment]
Hey there. I am currently experiencing this issue also on the WebKit build. Is there any update on progress for this/any pointers to where I could look into making a fix myself?

**Comment by imalsogreg** [comment]
@santiweight I have been under the gun at work and haven't made any progress on this. I would absolutely love help on it if you're willing to take a look! Most of my XHR handling either relies on low-level functions from `reflex-dom`, or uses copies of implementations of functions from `reflex-dom`. I suspect the issue is somewhere around here: https://github.com/imalsogreg/servant-reflex/blob/master/src/Servant/Common/Req.hs#L222
